src/common/Throttle.h

```diff
@@ -330,4 +332,84 @@ class OrderedThrottle {
   uint32_t waiters = 0;
 };
 
+
+class TokenBucketThrottle {
+
+  class Bucket {
+    CephContext *cct;
+    const std::string name;
+    std::atomic<int64_t> remain = { 0 }, max = { 0 };
+    Mutex lock;
```

Comment:
Why does this require its own lock?

Reply:
To protect the remain and max. Yes, the remain and max are both atomic but we need to make sure a function atomic. e.g, function ```put(c)```:
```
if (remain + c <= max)
    remain += c;
```
If the remain is 2 and max is 5. There are two threads A and B are doing put(2). 
```
Thread A                                                                Thread B
if (remain + c <= max) <---- 2 + 2 < 5
                                                                         if (remain + c <= max) <---- 2 + 2 < 5
                                                                               remain += c     <-------remain = 4
   remain += c  <-------- remain = 6
```
Then we probably get a remain > max.

Reply:
... but since ```Bucket``` is only accessible via ```TokenBucketThrottle``` and it already has a lock, why not just have ```TokenBucketThrottle``` use its lock to protect its nested class?

Reply:
do you mean TokenBucketThrottle::m_blockers_lock, this is to protect the blockers_list. but not to protect fields in bucket. for example, the set_max() is not protected by m_blockers_lock, will be in a race condition if we don't have a Bucket::lock.

Reply:
```TokenBucketThrottle::set_max``` can easily take the lock before invoking the method in ```TokenBucketThrottle::Bucket::set_max``` -- and ```TokenBucketThrottle::m_blockers_lock``` can be renamed ```TokenBucketThrottle::m_lock```

Reply:
Hm.... okey, not bad. thanx.